src/models/bundle-descriptor-constraints.ts

```diff
@@ -0,0 +1,288 @@
+
+// Constraints
+
+const ENVIRONMENT_VARIABLE_CONSTRAINTS: UnionTypeConstraints<EnvironmentVariable> =
+  [
+    {
+      name: {
+        required: true,
+        type: 'string'
+      },
+      value: {
+        required: true,
+        type: 'string'
+      }
+    },
+    {
+      name: {
+        required: true,
+        type: 'string'
+      },
+      valueFrom: {
+        required: true,
+        items: {
+          secretKeyRef: {
+            required: true,
+            items: {
+              name: {
+                required: true,
+                type: 'string'
+              },
+              key: {
+                required: true,
+                type: 'string'
+              }
+            }
+          }
+        }
+      }
+    }
+  ]
+
+const API_CLAIMS_CONSTRAINTS: UnionTypeConstraints<
+  ApiClaim | ExternalApiClaim
+> = [
+  {
+    name: {
+      required: true,
+      type: 'string'
+    },
+    type: {
+      required: true,
+      type: 'string',
+      validators: [values([ApiType.Internal])]
+    },
+    serviceId: {
+      required: true,
+      type: 'string'
+    }
+  },
+  {
+    name: {
+      required: true,
+      type: 'string'
+    },
+    type: {
+      required: true,
+      type: 'string',
+      validators: [values([ApiType.External])]
+    },
+    serviceId: {
+      required: true,
+      type: 'string'
+    },
+    bundleId: {
+      required: true,
+      type: 'string'
+    }
+  }
+]
+
+const MICROSERVICE_CONSTRAINTS: ObjectConstraints<MicroService> = {
+  name: {
+    required: true,
+    type: 'string',
+    validators: [nameRegExpValidator]
+  },
+  stack: {
+    required: true,
+    type: 'string',
+    validators: [values(MicroServiceStack)]
+  },
+  deploymentBaseName: {
+    required: false,
+    type: 'string'
+  },
+  dbms: {
+    required: true,
+    type: 'string'
+  },
+  ingressPath: {
+    required: false,
+    type: 'string'
+  },
+  healthCheckPath: {
+    required: false,
+    type: 'string'
+  },
+  roles: {
+    isArray: true,
+    required: false,
+    type: 'string'
+  },
+  securityLevel: {
+    required: false,
+    type: 'string',
+    validators: [values(SecurityLevel)]
+  },
+  permissions: {
+    isArray: true,
+    required: false,
+    children: {
+      clientId: {
+        required: true,
+        type: 'string'
+      },
+      role: {
+        required: true,
+        type: 'string'
+      }
+    }
+  },
+  env: {
+    isArray: true,
+    required: false,
+    children: ENVIRONMENT_VARIABLE_CONSTRAINTS
+  }
+}
+
+const MICROFRONTEND_CONSTRAINTS: ObjectConstraints<MicroFrontend> = {
+  name: {
+    required: true,
+    type: 'string',
+    validators: [nameRegExpValidator]
+  },
+  code: {
+    required: false,
+    type: 'string'
+  },
+  stack: {
+    required: true,
+    type: 'string',
+    validators: [values(MicroFrontendStack)]
+  },
+  titles: {
+    required: true,
+    validators: [isMapOfStrings],
+    items: {}
+  },
+  publicFolder: {
+    required: true,
```

Comment:
I'm not sure this is required as we should have a default value for it. @raxkaynan ?